### PR TITLE
Save Model Inputs, Model Outputs, Gradients, Custom Tensors, Layer Inputs, Layer Outputs

### PR DESCRIPTION
**Pull request by NihalHarish**
### Description of changes:
- This PR allows users to save model inputs, outputs, gradients, custom tensors, layer inputs and outputs.
- Description of new functions introduced:
   - `should_save_global_step_for_mode`: this function is not used anywhere in the `smdebug` codebase but instead used in AWS TF. It allows us to determine if a step should be saved by smdebug.
   - `_get_model`: refactored existing code into this fn to reduce duplication
   - `save_custom_tensor`: New fn introduced to allow users to save arbitrary tensors. This feature is currently only supported on TF2.
   - `save_custom_tensor`: abstracts the generic logic to save tensors
   - `save_smdebug_logs` : scan TF2 logs and saves model inputs, outputs, gradients, custom tensors, layer inputs and outputs
   -  `_save_layer_input_and_outputs` : logic to save layer inputs and outputs
   - `_wrap_model_with_input_output_saver`: attaches a hook to every layer of the model to allow smdebug to capture inputs and outputs.
   - `LayerWithHooks`: extension of the keras Layer class, to allow smdebug to capture Layer metrics.
   - `InputOutputSaver`: minimal hook that is attached to every layer to capture inputs and outputs
   - `get_layer_call_fn`: fn wrapper around the original Layer object's `__call__` fn.

#### Style and formatting:

I have run `pre-commit install` to ensure that auto-formatting happens with every commit.

#### Issue number, if available

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: smdebug/tensorflow/keras.py [hunk 1]
```diff
@@ -584,14 +717,13 @@ def _on_any_batch_end(self, batch, mode, logs=None):
         if not is_tf_version_2x() or (is_tf_version_2x() and not tf.executing_eagerly()):
             self._remove_fetches_and_callbacks(mode)
 
+        self._save_tensors_post_step(batch, logs)
```
Comment (NihalHarish) [inline smdebug/tensorflow/keras.py:720]: Moved it to this position so that it is called before `self._prepare_non_layer_tensors()`.

File: examples/tensorflow2/scripts/tf2_save_metrics.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
+"""
+This file is temporary, for testing with 2.X.
```
Comment (vandanavk) [inline examples/tensorflow2/scripts/tf2_save_metrics.py:2]: so these files will be deleted just before merge?

File: smdebug/tensorflow/base_hook.py [hunk 3]
```diff
@@ -437,7 +437,7 @@ def _write_for_tensor(self, tensor_name, tensor_value, save_collections, tensor_
 
         # this tensor_name is tf tensor name, need to convert to export_name
         tensor_ref = self._get_tensor_ref(tensor_name, save_collections=save_collections)
-        if tensor_ref:
+        if tensor_ref is not None:
```
Comment (vandanavk) [inline smdebug/tensorflow/base_hook.py:440]: why dint the original if condition work?

File: smdebug/tensorflow/keras.py [hunk 4]
```diff
@@ -366,6 +385,90 @@ def _add_metric(self, metric_name, metric_value: tf.Tensor = None):
             coll.set_tensor_ref(TensorRef.from_non_graph_var(metric_name))
         self.tensor_to_collections[metric_name] = {coll}
 
+    def save_custom_tensor(self, tensor_name, tensor_value, collections_to_write):
```
Comment (vandanavk) [inline smdebug/tensorflow/keras.py:388]: so same tensor can be saved under multiple collections? and we are relying on the user to pass the right arguments?

could you add the doc stub for this?

is this applicable only to TF?

if this is a user facing API, you will have to update docs.

File: smdebug/tensorflow/keras.py [hunk 5]
```diff
@@ -366,6 +385,90 @@ def _add_metric(self, metric_name, metric_value: tf.Tensor = None):
...
+        for collection in collections_to_write:
+            self.custom_tensors_to_save[tensor_name] = (tensor_value, collection)
+
+    def _save_custom_tensors_post_step(self):
```
Comment (vandanavk) [inline smdebug/tensorflow/keras.py:392]: description of what this function is for

File: smdebug/tensorflow/keras.py [hunk 6]
```diff
@@ -366,6 +385,90 @@ def _add_metric(self, metric_name, metric_value: tf.Tensor = None):
...
+        for tensor_name in self.custom_tensors_to_save:
+            tensor_value, collection_names = self.custom_tensors_to_save[tensor_name]
+            self._save_tensor(tensor_name, tensor_value, collection_names)
```
Comment (vandanavk) [inline smdebug/tensorflow/keras.py:395]: clear self.custom_tensors_to_save after you save them?

File: smdebug/tensorflow/keras.py [hunk 7]
```diff
@@ -366,6 +385,90 @@ def _add_metric(self, metric_name, metric_value: tf.Tensor = None):
...
+                    tensors_to_save = []
+                    export_name = get_model_output_export_name(key)
+                    tensors_to_save.append((export_name, logs[key]))
+                    collections_to_write = (
+                        {self.get_collection(CollectionKeys.OUTPUTS)}
+                        if self._is_collection_being_saved_for_step(CollectionKeys.OUTPUTS)
+                        else set()
+                    )
+                    for t_name, t_value in tensors_to_save:
+                        self._save_tensor(t_name, t_value, collections_to_write)
```
Comment (vandanavk) [inline smdebug/tensorflow/keras.py:439]: these 2 lines are in every if else condition except SMDEBUG_LAYER_OUTPUTS_KEY. can this be made common and be called just before function ends?

File: smdebug/tensorflow/utils.py [hunk 8]
```diff
@@ -261,6 +297,53 @@ def get_keras_layer_inputs(layer):
...
+
+    def __call__(self, callable_inputs, *args, **kwargs) -> None:
+        self.layer_input = kwargs["layer_input"]
+        self.layer_output = kwargs["layer_output"]
+
+
+def get_layer_call_fn(layer: tf.keras.layers.Layer) -> Callable[[tf.Tensor], tf.Tensor]:
+    old_call_fn = layer.call
+
+    def call(callable_inputs, *args, **kwargs) -> tf.Tensor:
```
Comment (vandanavk) [inline smdebug/tensorflow/utils.py:335]: when I was implementing wrap tape, I faced this problem where the tape was getting recursively wrapped. (thats why theres a unwrap_tape). is there a chance something similar will happen here?

File: tests/tensorflow2/test_keras.py [hunk 9]
```diff
@@ -37,7 +37,7 @@ def helper_keras_fit(
     eager=True,
     steps=None,
     add_callbacks=None,
-    run_eagerly=None,
+    run_eagerly=False,
```
Comment (vandanavk) [inline tests/tensorflow2/test_keras.py:40]: is this meant for model.compile()? where is it being used?

File: tests/tensorflow2/test_keras.py [hunk 10]
```diff
@@ -405,15 +405,17 @@ def test_keras_fit(out_dir, tf_eager_mode, saveall):
     helper_keras_fit(
         trial_dir=out_dir,
         hook=hook,
-        eager=tf_eager_mode,
+        run_eagerly=tf_eager_mode,
```
Comment (vandanavk) [inline tests/tensorflow2/test_keras.py:408]: the param "eager" is required for L43 above.

File: tests/tensorflow2/test_keras.py [hunk 11]
```diff
@@ -694,3 +772,34 @@ def input_fn():
     assert len(tr.steps()) == 2
     assert len(tr.steps(smd.modes.TRAIN)) == 1
     assert len(tr.steps(smd.modes.EVAL)) == 1
+
+
+@pytest.mark.skip
```
Comment (vandanavk) [inline tests/tensorflow2/test_keras.py:777]: why skipping?

File: tests/tensorflow2/test_keras_mirrored.py [hunk 12]
```diff
@@ -424,6 +424,7 @@ def test_clash_with_tb_callback(out_dir):
     assert len(tr.tensor_names()) == (10 if is_tf_2_2() else 11)
 
 
+@pytest.mark.skip
```
Comment (vandanavk) [inline tests/tensorflow2/test_keras_mirrored.py:427]: why?

File: examples/tensorflow2/scripts/tf2_save_metrics.py (hunk 2 above)
Comment (NihalHarish) [reply examples/tensorflow2/scripts/tf2_save_metrics.py:2]: These files are currently run on the AWS TF test pipeline.
They will be either modified or deleted after merging.

File: smdebug/tensorflow/base_hook.py (hunk 3 above)
Comment (NihalHarish) [reply smdebug/tensorflow/base_hook.py:440]: `if obj` executes `obj.__nonzero__()`
where `if obj is not None:` checks if it is not None.

This is  a nit fix.

File: tests/tensorflow2/test_keras.py (hunk 11 above)
Comment (NihalHarish) [reply tests/tensorflow2/test_keras.py:777]: Skipping it here because the github CI runs with a DLC that does not have my TF changes.

File: examples/tensorflow2/scripts/tf_save_metrics_gradient_tape.py [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+            # By default, the resources held by a GradientTape are released as
+            # soon as GradientTape.gradient() method is called. To compute
+            # multiple gradients over the same computation, create a persistent
+            # gradient tape. This allows multiple calls to the gradient() method
+            # as resources are released when the tape object is garbage collected.
+            if persistent:
+                _ = tape.gradient(loss_value, model.variables)
+            opt.apply_gradients(zip(grads, model.variables))
+            acc = train_acc_metric(dataset_labels, logits)
+            hook.record_tensor_value(tensor_name="accuracy", tensor_value=acc)
```
Comment (NihalHarish) [inline examples/tensorflow2/scripts/tf_save_metrics_gradient_tape.py:102]: Remove this API and merge with `save_tensor`

File: tests/tensorflow2/test_keras.py [hunk 14]
```diff
@@ -646,10 +654,80 @@ def test_keras_fit_pure_eager(out_dir, tf_eager_mode):
     helper_keras_fit(trial_dir=out_dir, hook=hook, eager=tf_eager_mode, run_eagerly=True)
 
     trial = smd.create_trial(path=out_dir)
-    assert len(trial.tensor_names()) == (12 if is_tf_2_2() else 13)
+    assert len(trial.tensor_names()) == (20 if is_tf_2_2() else 21)
```
Comment (NihalHarish) [inline tests/tensorflow2/test_keras.py:657]: save_tensor tensors should not be duplicated

File: smdebug/tensorflow/keras.py (hunk 6 above)
Comment (NihalHarish) [reply smdebug/tensorflow/keras.py:399]: Done.

File: smdebug/tensorflow/keras.py (hunk 7 above)
Comment (NihalHarish) [reply smdebug/tensorflow/keras.py:447]: Done

File: smdebug/tensorflow/utils.py (hunk 8 above)
Comment (NihalHarish) [reply smdebug/tensorflow/utils.py:335]: This is prevented by saving state to `self.has_registered_model`

File: examples/tensorflow2/scripts/tf_save_metrics_gradient_tape.py [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+
+    n_epochs = 1
+    for epoch in range(n_epochs):
+        for data, labels in dataset:
+            dataset_labels = labels
+            labels = tf.one_hot(labels, depth=10)
+            with hook.wrap_tape(tf.GradientTape(persistent=persistent)) as tape:
+                logits = model(data, training=True)
+                loss_value = cce(labels, logits)
+            hook.save_custom_tensor("y_labels", labels, "outputs")
```
[the remark this answers is not in the document]
Comment (NihalHarish) [reply examples/tensorflow2/scripts/tf_save_metrics_gradient_tape.py:90]: Done.

File: tests/tensorflow2/test_keras.py (hunk 10 above)
Comment (NihalHarish) [reply tests/tensorflow2/test_keras.py:408]: Yes

File: smdebug/tensorflow/keras.py [hunk 16]
```diff
@@ -95,6 +105,16 @@ def _is_not_supported(self):
                 self._hook_supported = False
         return not self._hook_supported
 
+    def should_save_global_step_for_mode(self, mode: str):
```
[the remark this answers is not in the document]
Comment (NihalHarish) [reply smdebug/tensorflow/keras.py:111]: Added comments below

File: smdebug/tensorflow/keras.py [hunk 17]
```diff
@@ -95,6 +105,16 @@ def _is_not_supported(self):
...
+        mode = str_to_mode_keys(mode)
+        mode_step = self.mode_steps[mode]
+        return self.save_config.should_save_step(mode, mode_step)
+
+    def register_model(self, model):
```
[the remark this answers is not in the document]
Comment (NihalHarish) [reply smdebug/tensorflow/keras.py:119]: Added comments below

File: smdebug/tensorflow/keras.py (hunk 5 above)
Comment (NihalHarish) [reply smdebug/tensorflow/keras.py:411]: Done.

File: smdebug/tensorflow/keras.py [hunk 18]
```diff
@@ -388,10 +497,38 @@ def _save_metrics(self, batch, logs, force_save=False):
                     self._add_metric(metric_name=key)
                     self._save_for_tensor(key, logs[key], check_before_write=False)
 
+    def _save_layer_input_and_outputs(self, grad_tape=False):
```
[the remark this answers is not in the document]
Comment (NihalHarish) [reply smdebug/tensorflow/keras.py:515]: commented below

File: smdebug/tensorflow/utils.py [hunk 19]
```diff
@@ -241,9 +273,13 @@ def is_keras_optimizer(obj):
     return False
 
 
-def get_export_name_for_keras(layer, tensor_type, tensor):
+def get_export_name_for_keras(layer, tensor_type, tensor=None):
     if tensor_type in ["input", "output", "weight"]:
-        return f"{layer.name}/{tensor_type}s/{tensor.name}"
+        if isinstance(layer, str):
+            # Tensor.name is meaningless when eager execution is enabled.
```
[the remark this answers is not in the document]
Comment (NihalHarish) [reply smdebug/tensorflow/utils.py:279]: See comment

File: tests/tensorflow2/test_keras.py (hunk 9 above)
Comment (NihalHarish) [reply tests/tensorflow2/test_keras.py:40]: yes

File: smdebug/tensorflow/keras.py [hunk 20]
```diff
@@ -366,6 +394,106 @@ def _add_metric(self, metric_name, metric_value: tf.Tensor = None):
             coll.set_tensor_ref(TensorRef.from_non_graph_var(metric_name))
         self.tensor_to_collections[metric_name] = {coll}
 
+    def save_tensor(self, tensor_name, tensor_value, collections_to_write=None):
+        if (
+            not ((isinstance(tensor_value, tf.Tensor)) and hasattr(tensor_value, "numpy"))
+        ) or self._is_not_supported():
+            return
+        if collections_to_write is None:
+            collections_to_write = "default"
```
Comment (vandanavk) [inline smdebug/tensorflow/keras.py:403]: nit: the default value on L397 can be "default" instead of this check?

File: smdebug/tensorflow/keras.py [hunk 21]
```diff
@@ -366,6 +394,106 @@ def _add_metric(self, metric_name, metric_value: tf.Tensor = None):
             coll.set_tensor_ref(TensorRef.from_non_graph_var(metric_name))
         self.tensor_to_collections[metric_name] = {coll}
 
+    def save_tensor(self, tensor_name, tensor_value, collections_to_write=None):
+        if (
+            not ((isinstance(tensor_value, tf.Tensor)) and hasattr(tensor_value, "numpy"))
```
Comment (vandanavk) [inline smdebug/tensorflow/keras.py:399]: do we need to check that its tf2 or is this applicable and tested with tf1 as well?

File: tests/tensorflow2/test_keras.py (hunk 10 above)
Comment (vandanavk) [reply tests/tensorflow2/test_keras.py:410]: so you should pass eager too?

File: smdebug/tensorflow/keras.py (hunk 20 above)
Comment (NihalHarish) [reply smdebug/tensorflow/keras.py:403]: Nice! Thanks!

File: tests/tensorflow2/test_keras.py (hunk 10 above)
Comment (NihalHarish) [reply tests/tensorflow2/test_keras.py:410]: `eager` is set to True by default.
`run_eagerly` indicates to the `model.compile` that the model should be run eagerly
